common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchLogStore.scala

```diff
@@ -0,0 +1,140 @@
+                                       actionField: String,
+                                       requiredHeaders: String = "")
+
+/**
+ * ElasticSearch based implementation of a DockerToActivationFileLogStore. When using the JSON log driver, docker writes
+ * stdout/stderr to JSON formatted files. Those files can be processed by a backend service asynchronously to store
+ * user logs in ElasticSearch. This log store allows user logs then to be fetched from ElasticSearch.
+ */
+class ElasticSearchLogStore(
+  system: ActorSystem,
+  httpFlow: Option[Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None,
+  destinationDirectory: Path = Paths.get("logs"),
+  elasticSearchConfig: ElasticSearchLogStoreConfig =
+    loadConfigOrThrow[ElasticSearchLogStoreConfig](ConfigKeys.elasticSearch))
+    extends DockerToActivationFileLogStore(system, destinationDirectory) {
+
+  // Schema of logs in ES
+  case class UserLogEntry(message: String, stream: String, time: String, action: String)
+
+  object UserLogEntry extends DefaultJsonProtocol {
+    implicit val serdes =
+      jsonFormat(
+        UserLogEntry.apply,
+        "message",
+        elasticSearchConfig.streamField,
+        "time_date",
+        elasticSearchConfig.actionField)
+  }
+
+  implicit val actorSystem = system
+
+  private val esClient = new ElasticSearchRestClient(
+    elasticSearchConfig.protocol,
+    elasticSearchConfig.host,
+    elasticSearchConfig.port,
+    httpFlow)
+  private val requiredHeaders = elasticSearchConfig.requiredHeaders match {
+    case headers if !headers.isEmpty => headers.split(",")
+    case _                           => Array[String]()
+  }
+  private val logQuery =
+    s"_type: ${elasticSearchConfig.logMessageField} AND ${elasticSearchConfig.activationIdField}: %s"
+
+  private def transcribeLogs(queryResult: EsSearchResult): ActivationLogs = {
+    val logs = queryResult.hits.hits.map(hit => {
+      val userLogEntry = hit.source.convertTo[UserLogEntry]
+      s"${userLogEntry.time} ${userLogEntry.stream}: ${userLogEntry.message.stripLineEnd}"
+    })
+
+    ActivationLogs(logs)
+  }
+
+  private def extractRequiredHeaders(request: HttpRequest) =
+    request.headers.filter {
+      case header: HttpHeader if requiredHeaders.contains(header.lowercaseName) => true
+      case _                                                                    => false
+    }.toList
+
+  private def generatePayload(activation: WhiskActivation) = {
+    val queryString = EsQueryString(logQuery.format(activation.activationId))
+    val queryOrder = EsQueryOrder("time_date", EsOrderAsc)
+
+    EsQuery(queryString, Some(queryOrder)).toJson.asJsObject
+  }
+
+  private def generatePath(user: Identity) = {
+    Uri(
+      elasticSearchConfig.path
+        .replace("$UUID", user.uuid.asString)
+        .replace("$DATE", LocalDate.now.format(DateTimeFormatter.ofPattern("yyyy.MM.dd"))))
```

Comment:
- Please cache `DateTimeFormatter.ofPattern("yyyy.MM.dd")`
- `replace` will create new Regex Matchers and run them each time this is called. Should we rather pass in an interpolated string and use the native `format` like